pkg/logs/input/docker/launcher.go

```diff
@@ -56,8 +59,22 @@ func NewLauncher(sources *config.LogSources, services *service.Services, pipelin
 		erroredContainerID: make(chan string),
 		lock:               &sync.Mutex{},
 	}
-	err := launcher.setup()
-	if err != nil {
+
+	var retryStrategy retry.Strategy
+	if shouldRetry {
+		retryStrategy = retry.RetryCount
+	} else {
+		retryStrategy = retry.OneTry
+	}
+	launcher.initRetry.SetupRetrier(&retry.Config{
+		Name:          "docker Launcher setup",
+		AttemptMethod: func() error { return launcher.setup() },
+		Strategy:      retryStrategy,
+		RetryCount:    math.MaxInt32,
+		RetryDelay:    30 * time.Second,
+	})
+
+	if err := launcher.initRetry.TriggerRetry(); err != nil && err.RetryStatus == retry.PermaFail {
 		return nil, err
 	}
 	// Sources and services are added after the setup to avoid creating
```

Comment:
I tried to keep that block stuck to `launcher.setup()`, so, inside the `AttemptMethod` of the retrier to guarantee that those actions are done after a successful call to `launcher.setup()` but it is not working.

Indeed, that’s the call to `sources.GetAddedForType(…)` that creates the channel if it is not existing.
The issue if that is done inside the retrier is that the `AddSource` is called before the channel is created so that, when the docker daemon is started, the `func (docker.Launcher *)run()` event loop receives a notification for the service, but never for the source and it cannot match the corresponding source for the new service.